File: app/src/test/java/com/nicholasworkshop/tinklabstest/fragment/guide/GuidePresenterTest.java
package com.nicholasworkshop.tinklabstest.fragment.guide;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import io.reactivex.Observable;

import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

/**
 * Created by nicholas.wong on 2017/05/17.
 */
public class GuidePresenterTest {

    @GuideModule.GuideType private static final int GUIDE_TYPE = 213;

    @Mock Context context;
    @Mock GuideView guideView;
    @Mock GuideModel guideModel;
    @Mock LayoutInflater layoutInflater;
    @Mock ViewGroup viewGroup;

    private GuidePresenter guidePresenter;

    @Before
    @SuppressWarnings("WrongConstant")
    public void setUp() throws Exception {
        initMocks(this);
        when(guideModel.storyList(anyInt(), anyInt())).thenReturn(Observable.empty());
        when(guideModel.ads()).thenReturn(Observable.empty());
        when(guideView.storyRequests()).thenReturn(Observable.empty());
        guidePresenter = new GuidePresenter(context, guideView, guideModel);
    }

    @Test
    public void createView() throws Exception {
        guidePresenter.createView(layoutInflater, viewGroup);
        verify(guideView).createView(layoutInflater, viewGroup);
    }

    @Test
    @SuppressWarnings("WrongConstant")
    public void viewCreated() throws Exception {
        guidePresenter.viewCreated(GUIDE_TYPE);
        verify(guideView).viewCreated(viewGroup);
        verify(guideModel).storyList(eq(GUIDE_TYPE), anyInt());
        verify(guideView).storyRequests();
        verify(guideModel).ads();
    }
}

File: app/src/test/kotlin/com/nicholasworkshop/tinklabstest/fragment/guide/GuidePresenterTest.kt
package com.nicholasworkshop.tinklabstest.fragment.guide

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Observable
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations.initMocks

class GuidePresenterTest {

    @Mock private lateinit var context: Context
    @Mock private lateinit var guideView: GuideView
    @Mock private lateinit var guideModel: GuideModel
    @Mock private lateinit var layoutInflater: LayoutInflater
    @Mock private lateinit var viewGroup: ViewGroup
    @Mock private lateinit var guideType: GuideType

    private lateinit var guidePresenter: GuidePresenter

    @Before
    fun setUp() {
        initMocks(this)
        whenever(guideModel.storyList(any(), anyInt())).thenReturn(Observable.empty())
        whenever(guideModel.ads()).thenReturn(Observable.empty())
        whenever(guideView.storyRequests()).thenReturn(Observable.empty())
        guidePresenter = GuidePresenter(context, guideView, guideModel)
    }

    @Test
    fun createView() {
        guidePresenter.createView(layoutInflater, viewGroup)
        verify(guideView).createView(layoutInflater, viewGroup)
    }

    @Test
    fun viewCreated() {
        guidePresenter.viewCreated(guideType)
        verify(guideView).viewCreated()
        verify(guideModel).storyList(eq(guideType), anyInt())
        verify(guideView).storyRequests()
        verify(guideModel).ads()
    }
}